frontend/packages/console-app/src/components/detect-namespace/namespace.ts

```diff
@@ -3,90 +3,89 @@ import * as React from 'react';
+  const [activeNamespace, setActiveNamespace] = React.useState<string>('');
   const dispatch = useDispatch();
   const setNamespace = React.useCallback(
     (namespace: string) => {
-      dispatch(setActiveNamespace(namespace));
+      setActiveNamespace(namespace);
+      dispatch(setActiveNamespaceForStore(namespace));
       setLastNamespace(namespace);
     },
     [dispatch, setLastNamespace],
   );
-  const useProjects = useFlag(FLAGS.OPENSHIFT);
+  const useProjects: boolean = useFlag(FLAGS.OPENSHIFT);
+
+  const resourcesLoaded: boolean = !!(
+    urlNamespace ||
+    (preferredNamespaceLoaded && lastNamespaceLoaded)
+  );
 
-  // Keep namespace in sync with redux.
   React.useEffect(() => {
-    // Url namespace overrides everything. We don't need to wait for loaded
-    if (urlNamespace) {
-      dispatch(setActiveNamespace(urlNamespace));
-    }
-    // Automatically sets favorited or latest namespace as soon as
-    // both informations are loaded from user settings.
-    if (!urlNamespace && favoriteLoaded && lastNamespaceLoaded && !flagPending(useProjects)) {
-      const setActiveIfExists = (ns: string) => {
-        if (!ns) {
-          return Promise.reject();
-        }
-        return namespaceExists(useProjects ? ProjectModel : NamespaceModel, ns)
-          .then(() => dispatch(setActiveNamespace(ns)))
-          .catch((res) => {
-            // eslint-disable-next-line no-console
-            console.warn('Error fetching namespace', ns);
-            throw res;
-          });
-      };
-      setActiveIfExists(favoritedNamespace)
-        .catch(() => setActiveIfExists(lastNamespace))
-        .catch(() => {});
+    if (resourcesLoaded) {
+      getValueForNamespace(
+        useProjects,
+        urlNamespace,
+        activeNamespace,
+        preferredNamespace,
+        lastNamespace,
+      )
+        .then((ns) => {
+          if (ns !== activeNamespace) {
+            setActiveNamespace(ns);
+            // sync with redux store
+            dispatch(setActiveNamespaceForStore(ns));
+          }
+        })
+        .catch((e) => {
+          // eslint-disable-next-line no-console
+          console.warn('Error fetching namespace', e);
+        });
     }
-    // Only run this hook after favorite and last namespace are loaded.
+    // Only run this hook after all resources have loaded.
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [favoriteLoaded, lastNamespaceLoaded, useProjects]);
+  }, [resourcesLoaded, useProjects]);
+
+  React.useEffect(() => {
+    if (urlNamespace) {
+      getValueForNamespace(useProjects, urlNamespace)
+        .then((ns) => {
+          setActiveNamespace(ns);
+          // sync with redux store
+          dispatch(setActiveNamespaceForStore(ns));
+        })
+        .catch((e) => {
+          // eslint-disable-next-line no-console
+          console.warn('Error fetching namespace', e);
+        });
+    }
+  }, [dispatch, urlNamespace, useProjects]);
```

Comment:
Won't this effect race against the other effect?

Reply:
@christianvogt Did you think it is necessary to check if the `urlNamespace` exists? For me it would be okay to show an "Project not found" error in this case instead of magically redirect the user when the user opens a specific namespaced link..

If this is okay we could use urlNamespace as default value for the activeNamespace state and don't need a second useEffect.

Reply:
The current implementation does the redirect and therefore I'm reluctant to change the behavior at this time.

Can we combine both effects together like it was done before?

Reply:
As far as I could understand we would still need a `useEffect`, since router links don't reload the console, so the state will not be reinitialized with the urlNamespace.
We can add a condition to avoid a race condition.
something like 
` React.useEffect(() => {
if (activeNamespace && urlNamespace && urlNamespace !== activeNamespace) {
// update state
}
`
WDYT?